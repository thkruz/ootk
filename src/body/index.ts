export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
