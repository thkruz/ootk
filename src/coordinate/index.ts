export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
