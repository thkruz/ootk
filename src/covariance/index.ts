export * from './CovarianceSample.js';
export * from './StateCovariance.js';
