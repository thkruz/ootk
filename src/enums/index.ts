export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
