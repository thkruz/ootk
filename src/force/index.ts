export { EarthGravity } from './EarthGravity.js';
export { Force } from './Force.js';
export { ForceModel } from './ForceModel.js';
export { Gravity } from './Gravity.js';
export { SolarRadiationPressure } from './SolarRadiationPressure.js';
export { ThirdBodyGravity } from './ThirdBodyGravity.js';
export { Thrust } from './Thrust.js';

