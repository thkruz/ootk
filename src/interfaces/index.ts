export type { BaseObjectParams } from './BaseObjectParams.js';
export type { GroundPositionParams } from './GroundPositionParams.js';
export type { OptionsParams } from './OptionsParams.js';
export type { SatelliteParams } from './SatelliteParams.js';
export type { SensorParams } from './SensorParams.js';
export type { StarObjectParams } from './StarObjectParams.js';
export type { ClassicalElementsParams } from './ClassicalElementsParams.js';
export type { EquinoctialElementsParams } from './EquinoctialElementsParams.js';
export type { RfSensorParams } from './RfSensorParams.js';
