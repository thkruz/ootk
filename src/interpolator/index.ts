export { ChebyshevCoefficients } from './ChebyshevCoefficients.js';
export { ChebyshevInterpolator } from './ChebyshevInterpolator.js';
export { CubicSpline } from './CubicSpline.js';
export { CubicSplineInterpolator } from './CubicSplineInterpolator.js';
export { Interpolator } from './Interpolator.js';
export { LagrangeInterpolator } from './LagrangeInterpolator.js';
export { StateInterpolator } from './StateInterpolator.js';
export { VerletBlendInterpolator } from './VerletBlendInterpolator.js';

