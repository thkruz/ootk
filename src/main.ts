/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
