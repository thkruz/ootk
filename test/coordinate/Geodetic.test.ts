import { Geodetic, AngularDistanceMethod, Degrees, EpochUTC, Kilometers, Radians } from '../../src/main';

describe('Geodetic', () => {
  /*
   * Creating a Geodetic object with valid latitude, longitude, and altitude
   * values should succeed.
   */
  it('should create a Geodetic object with valid latitude, longitude, and altitude values', () => {
    const geodetic = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);

    expect(geodetic).toMatchSnapshot();
  });

  // Converting a Geodetic object from degrees to radians should succeed.
  it('should convert a Geodetic object from degrees to radians', () => {
    const geodetic = Geodetic.fromDegrees(45 as Degrees, 90 as Degrees, 0 as Kilometers);

    expect(geodetic).toMatchSnapshot();
  });

  // Converting a Geodetic object to a string should succeed.
  it('should convert a Geodetic object to a string', () => {
    const geodetic = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const result = geodetic.toString();

    expect(result).toMatchSnapshot();
  });

  // Converting a Geodetic object to ITRF coordinates should succeed.
  it('should convert a Geodetic object to ITRF coordinates', () => {
    const geodetic = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const epoch = EpochUTC.fromDateTime(new Date('2024-01-14T14:39:39.914Z'));
    const result = geodetic.toITRF(epoch);

    expect(result).toMatchSnapshot();
  });

  /*
   * Calculating the angle between two Geodetic objects using Haversine method
   * should succeed.
   */
  it('should calculate the angle between two Geodetic objects using Haversine method', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 0 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 3.1015 as Radians, 0 as Kilometers);
    const result = geodetic1.angle(geodetic2);

    expect(result).toMatchSnapshot();
  });

  /*
   * Calculating the angle between two Geodetic objects using Spherical Law of
   * Cosines method should succeed.
   */
  it('should calculate the angle between two Geodetic objects using Spherical Law of Cosines method', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 0 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 3.1015 as Radians, 0 as Kilometers);
    const result = geodetic1.angle(geodetic2, AngularDistanceMethod.Cosine);

    expect(result).toMatchSnapshot();
  });

  /*
   * Creating a Geodetic object with invalid latitude value should raise an
   * exception.
   */
  it('should raise an exception when creating a Geodetic object with invalid latitude value', () => {
    expect(() => new Geodetic(100 as Radians, 1.5708 as Radians, 0 as Kilometers)).toThrow();
  });

  /*
   * Calculating the angle in degrees between two Geodetic objects using
   * Haversine method should succeed.
   */
  it('should calculate the angle in degrees between two Geodetic objects using Haversine method', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 3.1015 as Radians, 200 as Kilometers);
    const angleDegrees = geodetic1.angleDeg(geodetic2, AngularDistanceMethod.Haversine);

    expect(angleDegrees).toMatchSnapshot();
  });

  /*
   * Calculating the angle in degrees between two Geodetic objects using
   * Spherical Law of Cosines method should succeed.
   */
  it('should calculate the angle in degrees between two Geodetic objects using Spherical Law of Cosines method', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 3.1015 as Radians, 200 as Kilometers);
    const angleDegrees = geodetic1.angleDeg(geodetic2, AngularDistanceMethod.Cosine);

    expect(angleDegrees).toMatchSnapshot();
  });

  /*
   * Calculating the distance between two Geodetic objects using Haversine
   * method should succeed.
   */
  it('should calculate the distance between two Geodetic objects using Haversine method', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 0 as Kilometers);
    const geodetic2 = new Geodetic(0 as Radians, 0 as Radians, 0 as Kilometers);
    const distance = geodetic1.distance(geodetic2, AngularDistanceMethod.Haversine);

    expect(distance).toMatchSnapshot();
  });

  /*
   * Calculating the distance between two Geodetic objects using Spherical Law
   * of Cosines method should succeed.
   */
  it('should calculate the distance between two Geodetic objects using Spherical Law of Cosines method', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 0 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const distance = geodetic1.distance(geodetic2, AngularDistanceMethod.Cosine);

    expect(distance).toMatchSnapshot();
  });

  /*
   * Calculating the angle between two Geodetic objects with identical
   * coordinates should return 0 radians.
   */
  it('should return 0 radians when angle between two Geodetic objects with identical coordinates', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 200 as Kilometers);
    const angle = geodetic1.angle(geodetic2);

    expect(angle).toBe(0);
  });

  /*
   * Calculating the angle in degrees between two Geodetic objects with
   * identical coordinates should return 0 degrees.
   */
  it('should return 0 degrees when angle in degrees between two Geodetic objects with identical coordinates', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 200 as Kilometers);
    const angleDegrees = geodetic1.angleDeg(geodetic2);

    expect(angleDegrees).toBe(0);
  });

  /*
   * Calculating the distance between two Geodetic objects with identical
   * coordinates should return 0 kilometers.
   */
  it('should return 0 when calculating the distance between two Geodetic objects with identical coordinates', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const distance = geodetic1.distance(geodetic2);

    expect(distance).toBe(0);
  });

  /*
   * Checking if a Geodetic object is in view of another Geodetic object with
   * identical coordinates should return True.
   */
  it('should return true if a Geodetic object is in view of another Geodetic object with identical coordinates', () => {
    const geodetic1 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const geodetic2 = new Geodetic(0.7854 as Radians, 1.5708 as Radians, 100 as Kilometers);
    const result = geodetic1.isInView(geodetic2);

    expect(result).toBe(true);
  });
});
