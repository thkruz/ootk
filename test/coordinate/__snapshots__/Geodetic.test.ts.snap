// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geodetic should calculate the angle between two Geodetic objects using Haversine method 1`] = `1.0238932134660375`;

exports[`Geodetic should calculate the angle between two Geodetic objects using Spherical Law of Cosines method 1`] = `1.0238932134660375`;

exports[`Geodetic should calculate the angle in degrees between two Geodetic objects using Haversine method 1`] = `58.66475980369142`;

exports[`Geodetic should calculate the angle in degrees between two Geodetic objects using Spherical Law of Cosines method 1`] = `58.66475980369142`;

exports[`Geodetic should calculate the distance between two Geodetic objects using Haversine method 1`] = `10007.572625484478`;

exports[`Geodetic should calculate the distance between two Geodetic objects using Spherical Law of Cosines method 1`] = `0.0000949354182511178`;

exports[`Geodetic should convert a Geodetic object from degrees to radians 1`] = `
Geodetic {
  "alt": 0,
  "lat": 0.7853981633974483,
  "lon": 1.5707963267948966,
}
`;

exports[`Geodetic should convert a Geodetic object to ITRF coordinates 1`] = `
ITRF {
  "epoch": EpochUTC {
    "posix": 1705243179.914,
  },
  "position": Vector3D {
    "x": -0.01685374002219863,
    "y": 4588.292662103151,
    "z": 4558.066993511299,
  },
  "velocity": Vector3D {
    "x": 0,
    "y": 0,
    "z": 0,
  },
}
`;

exports[`Geodetic should convert a Geodetic object to a string 1`] = `
"Geodetic
  Latitude:  45.0001°
  Longitude: 90.0002°
  Altitude:  100.000 km"
`;

exports[`Geodetic should create a Geodetic object with valid latitude, longitude, and altitude values 1`] = `
Geodetic {
  "alt": 100,
  "lat": 0.7854,
  "lon": 1.5708,
}
`;
